src/compiler/scala/tools/nsc/Global.scala

```diff
@@ -385,45 +390,83 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
   abstract class GlobalPhase(prev: Phase) extends Phase(prev) {
     phaseWithId(id) = this
 
-    def run(): Unit = {
+    def apply(unit: CompilationUnit): Unit
+
+    def run() {
+      assertOnMainThread()
       echoPhaseSummary(this)
-      currentRun.units foreach applyPhase
+      Await.result(Future.sequence(currentRun.units map processUnit), Duration.Inf)
     }
 
-    def apply(unit: CompilationUnit): Unit
+    final def applyPhase(unit: CompilationUnit): Unit = Await.result(processUnit(unit), Duration.Inf)
 
-    /** Is current phase cancelled on this unit? */
-    def cancelled(unit: CompilationUnit) = {
-      // run the typer only if in `createJavadoc` mode
-      val maxJavaPhase = if (createJavadoc) currentRun.typerPhase.id else currentRun.namerPhase.id
-      reporter.cancelled || unit.isJava && this.id > maxJavaPhase
+    implicit val ec: ExecutionContext = {
+      val threadPoolFactory = ThreadPoolFactory(Global.this, this)
+      val javaExecutor = threadPoolFactory.newUnboundedQueueFixedThreadPool(parallelThreads, "worker")
+      scala.concurrent.ExecutionContext.fromExecutorService(javaExecutor, (_) => ())
     }
 
-    final def withCurrentUnit(unit: CompilationUnit)(task: => Unit): Unit = {
-      if ((unit ne null) && unit.exists)
-        lastSeenSourceFile = unit.source
-
+    private def processUnit(unit: CompilationUnit)(implicit ec: ExecutionContext): Future[Unit] = {
       if (settings.debug && (settings.verbose || currentRun.size < 5))
         inform("[running phase " + name + " on " + unit + "]")
-      if (!cancelled(unit)) {
-        currentRun.informUnitStarting(this, unit)
-        try withCurrentUnitNoLog(unit)(task)
-        finally currentRun.advanceUnit()
+
+      def runWithCurrentUnit(): Unit = {
+        val threadName = Thread.currentThread().getName
+        if (!threadName.contains("worker")) Thread.currentThread().setName(s"$threadName-worker")
+        val unit0 = currentUnit
+
+        try {
+          if ((unit ne null) && unit.exists) lastSeenSourceFile = unit.source
+          currentRun.currentUnit = unit
+          apply(unit)
+        } finally {
+          currentRun.currentUnit = unit0
+          currentRun.advanceUnit()
+          Thread.currentThread().setName(threadName)
+
+          // If we are on main thread it means there are no worker threads at all.
+          // That in turn means we were already using main reporter all the time, so there is nothing more to do.
+          // Otherwise we have to forward messages from worker thread reporter to main one.
+          reporter match {
```

Comment:
this doesn't work - we need to report in order - that was why I had the local data for this

we should move this to a relay method on the StoreReporter I think, or a new RelayingReporter or some such